### PR TITLE
[JENKINS-63539] Use additional repo URL variants to find cache

### PR DESCRIPTION
**Pull request by MarkEWaite**
## [JENKINS-63539](https://issues.jenkins-ci.org/browse/JENKINS-63539) Use additional repo URL variants to find cache

Estimation of repository size uses a cache key based on the repository URL.  There are several different forms of repository URL that can represent the same git repository.

For example, the following URL's all point to the same repository

* git://github.com/jenkinsci/git-plugin
* git@github.com:jenkinsci/git-plugin.git
* https://github.com/jenkinsci/git-plugin
* https://github.com/jenkinsci/git-plugin.git
* ssh://git@github.com/jenkinsci/git-plugin.git

This pull request permutes the repository URL provided by the user into several other forms in hopes that one of the forms already has a cached copy that can be used for local estimation of repository size.  This assumes that local Java processing of strings is much faster than a REST API call to the hosting provider to request the size of the repository.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] Documentation in README has been updated as necessary
- [x] Online help has been added and reviewed for any new or modified fields
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by rishabhBudhouliya — APPROVED** [review]
This is great, while creating `decideAndUseCache` I didn't think about the semantics of the URLs. Thanks for doing this, it is a great addition.

**Comment by MarkEWaite** [comment]
@rishabhBudhouliya I realized that I need more tests of the newly added method.  I'll add those tests before we merge this change.  The statement coverage of the functions looks reasonable but the number of assertions for those statements is far too low.

**Comment by MarkEWaite** [comment]
The tests lead me to #948 while trying to understand the branch coverage report.

**Comment by MarkEWaite** [comment]
@rishabhBudhouliya I reworked the commit to better match the cases that are being tested.  I believe the new implementation is simpler, clearer, and more accurate.  It also has many more tests than the first implementation I created.

File: src/main/java/jenkins/plugins/git/GitToolChooser.java [hunk 1]
```diff
@@ -153,17 +153,23 @@ private boolean decideAndUseCache(String remoteName) throws IOException, Interru
                 Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("jgit");
                 GitClient client = git.getClient();
                 if (client.hasGitRepo()) {
-                    sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
-                    sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+                    long clientRepoSize = FileUtils.sizeOfDirectory(cacheDir) / 1024;
+                    if (clientRepoSize > sizeOfRepo) {
```
Comment (rishabhBudhouliya) [inline src/main/java/jenkins/plugins/git/GitToolChooser.java:200]: In what event would we have multiple caches for the same git repository? 
Would it happen if two different multi branch projects work on a different version of the same git repository?

Comment (MarkEWaite) [reply src/main/java/jenkins/plugins/git/GitToolChooser.java:200]: I think that one case that might happen is two jobs that refer to a remote repository with different URLs and with different update frequencies.  If one of the caches is only updated once a week and the other is updated once an hour, the "once an hour" repository may have a larger size than the "once a week" repository.  This assumes the larger of the two values is the better approximation of repository size.

**Comment by MarkEWaite** [comment]
I've realized this is implemented incorrectly.  It needs rework to use much simpler code.

**Pull request closed without merging**